libnetwork/ipam/structures.go

```diff
@@ -37,8 +38,11 @@ type addrSpace struct {
 	subnets map[netip.Prefix]*PoolData
 
 	// Predefined pool for the address space
-	predefined           []netip.Prefix
-	predefinedStartIndex int
+	predefined map[ipamutils.IPVersion]*ipamutils.Subnetter
+
+	// reset indicates whether the predefined subnetters can be reset when all subnets
+	// have been enumerated but some have been released.
+	reset map[ipamutils.IPVersion]bool
```

Comment:
I still don't like this whole concept of holding the mutable iterator state with the collection being iterated over, and having this "can be reset" flag to paper over the issues stemming from that design. For one, there is no clean way to iterate over a permutation of the collection without repeating values. A concrete toy example might better illustrate what I'm trying to say.

Consider the following iterator, where `^` signifies the current position.
```
[ 1 2 3 4 5 ]
    ^
```
With an API similar to `Subnetter`, attempting to iterate over the whole collection starting from the current position would yield the sequence `2 3 4 5 1 2 3 4 5`. If one could assume the values in the collection are all unique, the iterating code could remember that `2` was the first value iterated over and stop once it iterates over `2` again. But that would leave the current position incremented one past where iteration "stopped," a fencepost error.

I think a better design would be to separate the iterator from the collection. Make the collection immutable and the iterator ephemeral, yielding exactly one permutation of the collection. Here's a sketch of what that could look like:
```go
// Cursor returns a cursor positioned at the start of the pool.
func (SubnetPool) Cursor() SubnetCursor

// Next returns a cursor moved forward by one position.
func (SubnetCursor) Advance() SubnetCursor
// Iter returns a new iterator over a permutation of the pool starting from the cursor position.
func (SubnetCursor) Iter() *SubnetIter

// Next advances the iterator cursor by one and returns the subnet at that position.
// If advancing the cursor would position it at a subnet
// which has already been returned by a prior call to Next,
// the cursor is not advanced and ok=false is returned.
func (*SubnetIter) Next() (v netip.Prefix, ok bool)
// Cursor returns a cursor for the current position.
func (SubnetIter) Cursor() SubnetCursor
```

The idea is to use the cursor in the same way as `predefinedStartIndex`: remember the last place iteration terminated so the next round of iteration knows where to start and stop.